src/ucs/datastruct/ptr_array.h

```diff
@@ -125,8 +124,100 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
  * Iterate over all valid elements in the array.
  */
 #define ucs_ptr_array_for_each(_var, _index, _ptr_array) \
-    for (_index = 0; _index < (_ptr_array)->size; ++_index) \
-         if (!__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index]))) \
+    unsigned elements = (_ptr_array)->n_occupied;\
+    unsigned is_element = 0;\
+    for (_index = 0; (elements > 0) && (_index < (_ptr_array)->size); elements -= is_element, ++_index)\
+       if ((is_element = ucs_likely(!__ucs_ptr_array_is_free(_var = (void *)((_ptr_array)->start[_index])))))\
```

Comment:
I think we can do it even faster. In the current code, if the size is 1000 and there's one element smack in the middle (and the rest is free) - this loop will do 500 iterations. why not use the space in the free elements to indicate where's the next occupied one?

Reply:
Yes, that's initially what I wanted to implement but then I figured the free-list is not necessarily sorted :-)
I.e. it is organized by the order of removed indices. This means that we cannot rely on the fact that freelist[i].index <= freelist[i+1].index and that all the elements between these indices are occupied... 
Storing the freelist sorted will cost O(n) on removal and on insert in the current form of the freelist [will require scanning the list to connect the chain properly].
Even if we decide we remove the PLACEHOLDER field from the free-list element and use these remaining 31 bits for storing the freelist->previous_index then we can implement a sorted freelist in O(1) for insert() but still, O(n) on removal [since it is unknown which free element is now required to be previous to the removed index and which one is required to be next].

Due to these considerations, I think the way the freelist remove is implemented today is good enough for now [I.e. freelist->head = removed_index, and removed_index->next=freelist->head ==> Thus, the head is replaced with the removed element in O(1)].

Besides this, when I added some code to store the first_occupied index element on every insert() operation [and does support the feature you mention], it slowed down the insert() by 5% - So we might want to implement another insert() API to also keep track of the first_occupied index to allow foreach to iterate from first_occupied_index to last_occupied_index...